server/scsynth/SC_PortAudio.cpp

```diff
@@ -88,7 +89,17 @@ class SC_PortAudioDriver : public SC_AudioDriver {
                           const PaStreamCallbackTimeInfo* timeInfo, PaStreamCallbackFlags statusFlags);
 
 private:
-    void GetPaDeviceFromName(const char* device, int* mInOut);
+    PaDeviceIndex GetPaDeviceFromName(const char* device, IOType ioType) const;
+    std::string GetDeviceName(int index) const;
+    // this function will select default PA devices if they are not defined
+    // it will also try to check for some configuration problems
+    // numIns, numOuts and sampleRate are only the requested values, can change later
+    PaError TryGetDefaultDevices(int* inDevice, int* outDevice, int numIns, int numOuts, double sampleRate) const;
+    template <typename SupportCheck> // expected signature: PaError (*)(PaStreamParameters&, double)
+    PaError CheckSinglePaDevice(int* device, double sampleRate, int maxChannels, int defaultDevice,
+                                const char* deviceType, SupportCheck isSupportedFunc) const;
+    void TryMatchDeviceSameAPI(int* matchingDevice, const int* knownDevice, IOType matchingDeviceType) const;
+    PaStreamParameters MakePaStreamParameters(int device, int channelCount, double suggestedLatency) const;
```

Comment:
These private member functions can all be declared `static` - as far as I can see they only operate on their arguments and don't access any instance member variables?

Reply:
These functions were `const` I've changed them to `static`. 

I have only a very rudimentary practical understanding when a function should be either `static` or `const` (I know these are two separate things, but AFAIU somewhat related). I just would like your confirmation this is the right choice, as I don't fully understand the difference.

Reply:
> I know these are two separate things, but AFAIU somewhat related

`static` member function - cannot access object state, basically an ordinary function, with access to private members of instances of the class of which they are declared a member
`const` member function - cannot mutate members (except for members declared `mutable`)

> I have only a very rudimentary practical understanding when a function should be either static or const

IMO, either is fine. `static` is nice because it lets you visually confirm that the function uses no object state. But in this case, if I were writing it, I wouldn't mind leaving them as const members. On the off chance that in the future these functions do need access to class state (and I judge that as fairly likely), it's kind of annoying to have to rework the interface by changing `static` back to `const`. In addition, these functions seem conceptually not `static` to me since they do indirectly access or modify class state.

Reply:
Thank you for the explanation. I reverted that commit so it's back to `const`.

Reply:
I don't want to argue with Brain or make you change this again, but just to clarify why I suggested it in the first place. The use of static here would convey that they are utility functions with a strong association to the class, that don't have direct access to members - which is the case. If the nature of the function(s) change later, then it would need re-writing, but given that they are just utility functions that query portaudio it seemed appropriate. It's fine either way really, so stick with Brian on this one :)

Reply:
to be clear, i am fine with either, i just personally would have written them const.